libsolidity/ast/Types.cpp

```diff
@@ -199,18 +148,78 @@ util::Result<TypePointers> transformParametersToExternal(TypePointers const& _pa
 	return transformed;
 }
 
+bool compareTypes(Type const* lhs, Type const* rhs)
+{
+	solAssert(lhs && rhs, "");
+	return lhs->richIdentifier() < rhs->richIdentifier();
 }
+using TypeSet = std::set<Type const*, decltype(compareTypes)*> ;
 
-vector<frontend::Type const*> solidity::frontend::oversizedSubtypes(frontend::Type const& _type)
+void storageOversizedSubtypesInner(
+	Type const* _type,
+	bool _includeType,
+	set<string>& _seen,
+	TypeSet& _oversizedSubtypes
+)
 {
-	set<StructDefinition const*> structsSeen;
-	TypeSet oversized;
-	oversizedSubtypesInner(_type, false, structsSeen, oversized);
-	vector<frontend::Type const*> res;
+	for (Type const* subtype: _type->subtypes())
+	{
+		string id = subtype->richIdentifier();
+		if (_seen.count(id))
+			continue;
+		_seen.insert(id);
+		bool includeType = _type->category() != Type::Category::Struct;
+		storageOversizedSubtypesInner(subtype, includeType, _seen, _oversizedSubtypes);
+		_seen.erase(id);
+	}
+
+	if (_includeType && _type->storageSizeUpperBound() >= bigint(1) << 64)
+		_oversizedSubtypes.insert(_type);
+}
+
+}
+
+vector<Type const*> Type::storageOversizedSubtypes() const
+{
+	set<string> seen;
+	TypeSet oversized(compareTypes);
+	storageOversizedSubtypesInner(this, false, seen, oversized);
+	vector<Type const*> res;
 	copy(oversized.cbegin(), oversized.cend(), back_inserter(res));
 	return res;
 }
 
+BoolResult Type::validForLocation(DataLocation _loc, set<string>* _seen) const
+{
+	if (_seen == nullptr)
+	{
+		set<string> seen;
+		return validForLocation(_loc, &seen);
+	}
+
+	for (Type const* subtype: subtypes())
+	{
+		string id = subtype->richIdentifier();
+		if (_seen->count(id))
+			continue;
+		_seen->insert(id);
+		BoolResult result = subtype->validForLocation(_loc, _seen);
+		_seen->erase(id);
+		if (!result)
+			return result;
+	}
+
+	bigint size = sizeUpperBound(_loc);
+	if (_loc == DataLocation::Storage && size >= bigint(1) << 256)
+		return BoolResult::err("Type too large for storage.");
+	if (_loc == DataLocation::Memory && size >= numeric_limits<unsigned>::max())
+		return BoolResult::err("Type too large for memory.");
+	if (_loc == DataLocation::CallData && size >= numeric_limits<unsigned>::max())
+		return BoolResult::err("Type too large for calldata.");
```

Comment:
It seems to me that this code expects `sizeUpperBound()` to return the number of slots in case of storage and the number of bytes in case of memory/calldata. I'm not sure if it's a good idea.

Reply:
Agreed. See above.